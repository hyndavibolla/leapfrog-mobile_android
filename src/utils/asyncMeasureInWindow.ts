import { View } from 'react-native';

import { IMeasure } from '../models/general';

export const asyncMeasureInWindow = (view?: View) =>
  new Promise<IMeasure>(resolve => {
    if (typeof view?.measureInWindow === 'function') return view.measureInWindow((x, y, width, height) => resolve({ x, y, width, height }));
    resolve({ x: 0, y: 0, width: 0, height: 0 });
  });
