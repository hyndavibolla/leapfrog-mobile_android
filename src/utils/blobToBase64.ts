/* istanbul ignore file */
// Ignoring for coverage: most types are native so do not require testing.
export const blobToBase64 = blob => {
  const reader = new FileReader();
  const resolved = new Promise(resolve => {
    reader.onloadend = () => {
      const dataUrl = reader.result as string;
      resolve(dataUrl.split(',')[1]);
    };
  });
  reader.readAsDataURL(blob);
  return resolved;
};
