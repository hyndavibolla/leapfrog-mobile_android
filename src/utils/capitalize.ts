export const capitalize = (str: string) => {
  str = str || '';
  return `${str.charAt(0).toUpperCase()}${str.substring(1, str.length).toLocaleLowerCase()}`;
};
