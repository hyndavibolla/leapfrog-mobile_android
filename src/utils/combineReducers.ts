import { IAction, IReducer } from '../models/general';

export function combineReducers<GS>(reducerMap: Record<string, (state: any, action: IAction) => any>): IReducer<GS> {
  return (gState: GS, gAction: IAction) => {
    const nextState = Object.entries(reducerMap).reduce(
      (total, [stateKey, reducer]) => ({ ...total, [stateKey]: reducer(gState[stateKey], gAction) }),
      {} as GS
    );
    const stateKeys = Array.from(new Set([...Object.keys(gState), ...Object.keys(nextState)]));
    const hasChanged = stateKeys.some(key => gState[key] !== nextState[key]);
    return hasChanged ? nextState : gState;
  };
}
