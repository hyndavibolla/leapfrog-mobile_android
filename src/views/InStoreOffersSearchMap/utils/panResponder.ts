const swipeConfig = {
  velocityThreshold: 0.3,
  directionalOffsetThreshold: 80,
  gestureIsClickThreshold: 5
};

const swipeDirections = {
  swipeUp: 'SWIPE_UP',
  swipeDown: 'SWIPE_DOWN'
};

export const handleShouldSetPanResponder = ({ nativeEvent: { touches } }, gestureState) => {
  return touches.length === 1 && !gestureIsClick(gestureState);
};

export const gestureIsClick = ({ dx, dy }) => {
  return Math.abs(dx) < swipeConfig.gestureIsClickThreshold && Math.abs(dy) < swipeConfig.gestureIsClickThreshold;
};

export const handlePanResponderEnd = (gestureState, onSwipeUp, onSwipeDown) => {
  const swipeDirection = getSwipeDirection(gestureState);
  triggerSwipeHandlers(swipeDirection, onSwipeUp, onSwipeDown);
};

export const triggerSwipeHandlers = (swipeDirection, onSwipeUp, onSwipeDown) => {
  const { swipeUp, swipeDown } = swipeDirections;
  switch (swipeDirection) {
    case swipeUp:
      onSwipeUp();
      break;
    case swipeDown:
      onSwipeDown();
      break;
  }
};

export const getSwipeDirection = gestureState => {
  const { swipeUp, swipeDown } = swipeDirections;
  if (isValidVerticalSwipe(gestureState)) {
    return gestureState.dy > 0 ? swipeDown : swipeUp;
  }
  return null;
};

export const isValidVerticalSwipe = ({ vy, dx }) => {
  const { velocityThreshold, directionalOffsetThreshold } = swipeConfig;
  return isValidSwipe(vy, velocityThreshold, dx, directionalOffsetThreshold);
};

export const isValidSwipe = (velocity, velocityThreshold, directionalOffset, directionalOffsetThreshold) => {
  return Math.abs(velocity) > velocityThreshold && Math.abs(directionalOffset) < directionalOffsetThreshold;
};
