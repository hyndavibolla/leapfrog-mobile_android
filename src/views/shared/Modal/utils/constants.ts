export enum ModalType {
  CENTER = 'center',
  BOTTOM = 'bottom'
}

export enum ModalSize {
  SMALL = 'small',
  MEDIUM = 'medium',
  EXTRA_LARGE = 'extralarge',
  DYNAMIC = 'dynamic',
  FULL_SCREEN = 'fullscreen'
}

export const ModalSizeToHeightMap = {
  [ModalSize.SMALL]: 40,
  [ModalSize.MEDIUM]: 50,
  [ModalSize.EXTRA_LARGE]: 80,
  [ModalSize.FULL_SCREEN]: 100
};
