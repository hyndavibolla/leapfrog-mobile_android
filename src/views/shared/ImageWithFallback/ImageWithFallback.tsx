import React, { memo, useCallback, useMemo, useState } from 'react';
import { ImageSourcePropType, ImageStyle, StyleProp } from 'react-native';
import Image, { FastImageProps } from 'react-native-fast-image';

import { useDebounce } from '../../../utils/useDebounce';
import { useTestingHelper } from '../../../utils/useTestingHelper';
import { ENV } from '../../../constants';

export interface Props extends FastImageProps {
  fallbackSource: ImageSourcePropType;
  fallbackStyle?: StyleProp<ImageStyle>;
}

export const ImageWithFallback = ({ fallbackSource, fallbackStyle, ...props }: Props) => {
  const { getTestIdProps } = useTestingHelper('image-with-fallback');
  const [error, setError] = useState(null);
  const [loading, setLoading] = useState(false);
  const source = useMemo(() => (error || !props.source ? fallbackSource : props.source), [error, props.source, fallbackSource]);
  const style = useMemo(() => [props.style, (error || !props.source) && fallbackStyle], [error, props.style, props.source, fallbackStyle]);
  const debouncedLoading = useDebounce(loading, ENV.IMAGE_LOADING_FLICKERING_DEBOUNCE_MS); // avoids flickering when image was cached
  const setLoadingStart = useCallback(() => setLoading(true), []);
  const setLoadingEnd = useCallback(() => setLoading(false), []);

  return (
    <>
      {!debouncedLoading ? null : <Image {...(props as any)} source={fallbackSource} style={fallbackStyle} {...getTestIdProps('loading')} />}
      <Image
        onLoadStart={setLoadingStart}
        onLoadEnd={setLoadingEnd}
        {...(props as any)}
        source={source}
        style={debouncedLoading ? { width: 0, height: 0 } : style}
        onError={setError}
        {...getTestIdProps('container')}
      />
    </>
  );
};

export default memo(ImageWithFallback);
