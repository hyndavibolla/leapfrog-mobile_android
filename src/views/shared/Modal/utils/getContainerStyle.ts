import { memorize } from '../../../../utils/memorize';
import { styles } from '../styles';
import { ModalSize, ModalType, ModalSizeToHeightMap } from './constants';

interface props {
  height: number;
  modalHeight: number;
  size: ModalSize;
  type: ModalType;
}

const getContainerStyle = ({ height, modalHeight, size, type }: props) => {
  switch (type) {
    case ModalType.BOTTOM:
      if (size === ModalSize.FULL_SCREEN) return [styles.containerBottom, styles.containerBottomFullScreen];
      else if (size !== ModalSize.DYNAMIC) return [styles.containerBottom, { maxHeight: (ModalSizeToHeightMap[size] * height) / 100 }];
      return [styles.containerBottom];
    case ModalType.CENTER:
      if (size === ModalSize.DYNAMIC) return [styles.containerCenter, styles.containerCenterDynamic, { bottom: (height - modalHeight) / 2 }];
      const center = (ModalSizeToHeightMap[size] * height) / 100;
      return [styles.containerCenter, { height: center, bottom: (height - center) / 2 }];
  }
};

export default memorize(getContainerStyle);
