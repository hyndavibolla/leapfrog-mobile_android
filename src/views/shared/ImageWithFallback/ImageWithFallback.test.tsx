import React from 'react';
import { fireEvent, render } from '@testing-library/react-native';
import { act } from 'react-test-renderer';
import { wait } from '../../../utils/wait';

import ImageWithFallback from './ImageWithFallback';

describe('ImageWithFallback', () => {
  it('should render', () => {
    const { toJSON } = render(<ImageWithFallback source={{ uri: 'uri' }} fallbackSource={{ uri: 'fallback' }} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('should render with fallback', () => {
    const { toJSON, getByTestId } = render(<ImageWithFallback source={{ uri: 'uri' }} fallbackSource={{ uri: 'fallback' }} />);
    fireEvent(getByTestId('image-with-fallback-container'), 'onError', new Error());
    expect(toJSON()).toMatchSnapshot();
  });

  it('should render with fallback style', () => {
    const { toJSON, getByTestId } = render(
      <ImageWithFallback source={{ uri: 'uri' }} fallbackSource={{ uri: 'fallback' }} fallbackStyle={{ backgroundColor: 'red' }} />
    );
    fireEvent(getByTestId('image-with-fallback-container'), 'onError', new Error());
    expect(toJSON()).toMatchSnapshot();
  });

  it('should render loading', async () => {
    const { queryByTestId, getByTestId } = render(
      <ImageWithFallback source={{ uri: 'uri' }} fallbackSource={{ uri: 'fallback' }} fallbackStyle={{ backgroundColor: 'red' }} />
    );
    await act(() => wait(0));
    fireEvent(getByTestId('image-with-fallback-container'), 'onLoadStart');
    await act(() => wait(0));
    expect(queryByTestId('image-with-fallback-loading')).toBeTruthy();
    fireEvent(getByTestId('image-with-fallback-container'), 'onLoadEnd');
    await act(() => wait(0));
    expect(queryByTestId('image-with-fallback-loading')).not.toBeTruthy();
  });
});
