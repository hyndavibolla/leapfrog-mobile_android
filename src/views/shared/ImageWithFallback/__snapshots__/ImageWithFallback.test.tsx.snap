// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageWithFallback should render 1`] = `
<View
  style={
    Array [
      Object {
        "overflow": "hidden",
      },
      Array [
        undefined,
        false,
      ],
    ]
  }
>
  <FastImageView
    accessibilityLabel="image-with-fallback-container"
    accessible={true}
    onFastImageError={[Function]}
    onFastImageLoadEnd={[Function]}
    onFastImageLoadStart={[Function]}
    resizeMode="cover"
    source={
      Object {
        "uri": "uri",
      }
    }
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
    testID="image-with-fallback-container"
  />
</View>
`;

exports[`ImageWithFallback should render with fallback 1`] = `
<View
  style={
    Array [
      Object {
        "overflow": "hidden",
      },
      Array [
        undefined,
        undefined,
      ],
    ]
  }
>
  <FastImageView
    accessibilityLabel="image-with-fallback-container"
    accessible={true}
    onFastImageError={[Function]}
    onFastImageLoadEnd={[Function]}
    onFastImageLoadStart={[Function]}
    resizeMode="cover"
    source={
      Object {
        "uri": "fallback",
      }
    }
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
    testID="image-with-fallback-container"
  />
</View>
`;

exports[`ImageWithFallback should render with fallback style 1`] = `
<View
  style={
    Array [
      Object {
        "overflow": "hidden",
      },
      Array [
        undefined,
        Object {
          "backgroundColor": "red",
        },
      ],
    ]
  }
>
  <FastImageView
    accessibilityLabel="image-with-fallback-container"
    accessible={true}
    onFastImageError={[Function]}
    onFastImageLoadEnd={[Function]}
    onFastImageLoadStart={[Function]}
    resizeMode="cover"
    source={
      Object {
        "uri": "fallback",
      }
    }
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
    testID="image-with-fallback-container"
  />
</View>
`;
